db_service/vcap.go

```diff
@@ -0,0 +1,98 @@
+package db_service
+
+import (
+	"code.cloudfoundry.org/lager"
+	"encoding/json"
+	"fmt"
+	"github.com/spf13/viper"
+	"net/url"
+	"os"
+)
+
+type VcapService struct {
+	BindingName  string            `json:"binding_name"`  // The name assigned to the service binding by the user.
+	InstanceName string            `json:"instance_name"` // The name assigned to the service instance by the user.
+	Name         string            `json:"name"`          // The binding_name if it exists; otherwise the instance_name.
+	Label        string            `json:"label"`         // The name of the service offering.
+	Tags         []string          `json:"tags"`          // An array of strings an app can use to identify a service instance.
+	Plan         string            `json:"plan"`          // The service plan selected when the service instance was created.
+	Credentials  map[string]string `json:"credentials"`   // The service-specific credentials needed to access the service instance.
+}
+
+func useVcapServices(logger lager.Logger) error {
+	vcapData, vcapExists := os.LookupEnv("VCAP_SERVICES")
+	if vcapExists {
+		vcapService, err := parseVcapServices(vcapData, logger)
+		if err != nil {
+			return fmt.Errorf("Error parsing VCAP_SERVICES: %s", err)
+		}
+
+		// if URI is supplied, we should parse it to fill any missing fields
+		u, err := url.Parse(vcapService.Credentials["uri"])
+		if err != nil {
+			return fmt.Errorf("Error parsing VCAP_SERVICES credentials URI: %s", err)
+		}
+
+		logger.Info("Using MySQL database injected via VCAP_SERVICES environment variable")
+		viper.Set(dbPathProp, u.Path)
+		viper.Set(dbTypeProp, DbTypeMysql)
+		viper.Set(dbHostProp, vcapService.Credentials["host"])
+		viper.Set(dbUserProp, vcapService.Credentials["Username"])
```

Comment:
for these, could we instead set the first of the following that exist: `url.Username(), vcapService.Credentials["Username"], vcapService.Credentials["username"]` ?